### PR TITLE
✨ Initial e2e test setup

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
kind/feature

**What this PR does / why we need it**:

This PR implements initial test suite setup for e2e testing. E2E tests are mostly based on helm charts of various versions. Rest of the test will be added later.

Example run using github actions: https://github.com/Danil-Grigorev/rancher-turtles/actions/runs/5904216470
Artifacts with collected logs from both workload and child clusters: https://github.com/Danil-Grigorev/rancher-turtles/suites/15233313000/artifacts/870563090
Imported rancher cluster from CAPI cluster within this run:
image

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Addresses #46 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [x] includes documentation
- [ ] adds unit tests
- [x] adds or updates e2e tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -360,6 +396,12 @@ $(KUSTOMIZE): # Build kustomize from tools folder.
 $(SETUP_ENVTEST): # Build setup-envtest from tools folder.
 	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) $(SETUP_ENVTEST_PKG) $(SETUP_ENVTEST_BIN) $(SETUP_ENVTEST_VER)
 
+$(GINKGO): # Build ginkgo from tools folder.
+	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) $(GINKGO_PKG) $(GINKGO_BIN) $(GINGKO_VER)
+
+$(CAPI_OPERATOR):
+	wget $(CAPI_OPERATOR_URL) -O $(CAPI_OPERATOR)
```
Comment (alexander-demicev) [inline Makefile:403]: do we download chat because it's installed from code in e2e setup?

File: test/e2e/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Test
+
+## Compatibility notice
+
+This package is not subject to deprecation notices or compatibility guarantees.
+
+- Breaking changes are likely. External providers using this package should update to the latest API changes when updating Cluster API Operator. Maintainers and contributors must give notice in release notes when a breaking change happens.
```
Comment (alexander-demicev) [inline test/e2e/README.md:7]: 👍 
we can always set depreciation annotation on methods, IDEs and linters will complain in that case

**Review by alexander-demicev — APPROVED** [review]
a good start, thanks a lot!

**Review by salasberryfin — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (Danil-Grigorev) [reply Makefile:403]: Yes, I’d like to add a way to override CAPI operator image with a custom chart, if there will be some major issues during testing we didn’t saw yet, but could be hot-fixed temporarily

File: test/e2e/e2e_suite_test.go [hunk 3]
```diff
@@ -260,6 +274,29 @@ func initBootstrapCluster(bootstrapClusterProxy framework.ClusterProxy, config *
...
+			"--create-namespace",
+			"--wait",
+		},
+	}
+	_, err = chart.InstallChart(map[string]string{
+		"bootstrapPassword":         "admin",
+		"features":                  "embedded-cluster-api=false",
+		"global.cattle.psp.enabled": "false",
+		"replicas":                  "1",
+		"hostname":                  os.Getenv("RANCHER_HOSTNAME"),
```
Comment (alexander-demicev) [inline test/e2e/e2e_suite_test.go:297]: This will have to be resolvable in test environment, you can do something similar to this https://github.com/rancher/eks-operator/blob/master/Makefile#L112 + sslip.io

**Review by furkatgofurov7 — COMMENTED** [review]
Great work, few suggestions:

File: test/e2e/resources/fleet-capi-test.yaml [hunk 4]
```diff
@@ -0,0 +1,4896 @@
+---
```
Comment (alexander-demicev) [inline test/e2e/resources/fleet-capi-test.yaml:1]: how is it related to fleet?

File: test/e2e/helpers_test.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"context"
+	_ "embed"
+)
+
+var (
+	ctx = context.Background()
+
+	//go:embed resources/fleet-capi-test.yaml
+	fleetCapiTestdata []byte
```
Comment (alexander-demicev) [inline test/e2e/helpers_test.go:31]:
```suggestion
	fleetCAPITestdata []byte
```

File: test/e2e/resources/fleet-capi-test.yaml [hunk 6]
```diff
@@ -0,0 +1,4896 @@
...
+      controlPlane: false
+    fd8:
+      controlPlane: false
+---
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+metadata:
+  labels:
+    cni: test1-crs-0
+    cluster-api.cattle.io/rancher-auto-import: "true"
```
Comment (alexander-demicev) [inline test/e2e/resources/fleet-capi-test.yaml:31]: can we set the label in the code instead? We want to test a case with a label on the namespace too, we can reuse this template.

File: test/e2e/import_test.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		}
+	})
+
+	It("should successfully create a rancher cluster from CAPI cluster", func() {
+		By("Creating a CAPI docker provider cluster setup")
+		Eventually(func() error {
+			return bootstrapClusterProxy.Apply(ctx, fleetCapiTestdata)
+		}, e2eConfig.GetIntervals(bootstrapClusterProxy.GetName(), "wait-controllers")...).Should(Succeed())
+
+		By("Waiting for rancher cluster record to appear")
```
Comment (alexander-demicev) [inline test/e2e/import_test.go:62]: should also check that it was imported succesfully

File: test/e2e/e2e_suite_test.go [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+	})
+	Expect(err).ToNot(HaveOccurred())
+}
+
+// Using a SynchronizedAfterSuite for controlling how to delete resources shared across ParallelNodes (~ginkgo threads).
+// The bootstrap cluster is shared across all the tests, so it should be deleted only after all ParallelNodes completes.
+var _ = SynchronizedAfterSuite(func() {
+	// After each ParallelNode.
+}, func() {
+	// After all ParallelNodes.
```
Comment (alexander-demicev) [inline test/e2e/e2e_suite_test.go:299]: we can dump artifacts here

File: test/e2e/helpers_test.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	//go:embed resources/docker-infra-secret.yaml
+	dockerVariablesSecret []byte
+)
+
+const (
+	operatorNamespace       = "capi-operator-system"
+	rancherTurtlesNamespace = "rancher-turtles-system"
+	rancherNamespace        = "cattle-system"
+	capiClusterName         = "test1"
+	capiClusterNamespace    = "fleet-default"
```
Comment (alexander-demicev) [inline test/e2e/helpers_test.go:42]: what is the reason for choosing fleet namespace?

File: Makefile [hunk 10]
```diff
@@ -113,7 +146,7 @@ GH_REPO_NAME ?= rancher-turtles
 GH_REPO ?= $(GH_ORG_NAME)/$(GH_REPO_NAME)
 
 # Allow overriding the imagePullPolicy
-PULL_POLICY ?= Always
+PULL_POLICY ?= IfNotPresent
```
Comment (alexander-demicev) [inline Makefile:149]: why was this changed?

File: test/e2e/e2e_suite_test.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+	operatorframework "sigs.k8s.io/cluster-api-operator/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/bootstrap"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	operaratorPackage = "CAPI_OPERATOR"
```
Comment (alexander-demicev) [inline test/e2e/e2e_suite_test.go:46]:
```suggestion
	operatorPackage = "CAPI_OPERATOR"
```

File: Makefile (hunk 10 above)
Comment (Danil-Grigorev) [reply Makefile:149]: This causes the cluster to pull the tag from outside. This should not be happening, e2e run should use the latest local image if it is specified. And it always is for rancher-turtles

File: test/e2e/resources/fleet-capi-test.yaml (hunk 6 above)
Comment (Danil-Grigorev) [reply test/e2e/resources/fleet-capi-test.yaml:31]: Yeah, makes sense, will update.

File: test/e2e/resources/fleet-capi-test.yaml (hunk 4 above)
Comment (Danil-Grigorev) [reply test/e2e/resources/fleet-capi-test.yaml:1]: https://github.com/richardcase/fleet-clusters

File: test/e2e/e2e_suite_test.go (hunk 3 above)
Comment (Danil-Grigorev) [reply test/e2e/e2e_suite_test.go]: Not entirely true. You can always manually open an ingress connection to the rancher cluster, and make something like localhost resolvable. Don’t want to overcomplicate it, it could be resolvable but it does not have to.

File: test/e2e/e2e_suite_test.go (hunk 8 above)
Comment (Danil-Grigorev) [reply test/e2e/e2e_suite_test.go:323]: Added correct machine log handling here. Really helpful, thanks

File: test/e2e/import_test.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
+//go:build e2e
+// +build e2e
+
+/*
+Copyright 2023 The Kubernetes Authors.
```
Comment (richardcase) [inline test/e2e/import_test.go:5]:
```suggestion
Copyright 2023 SUSE
```

File: test/e2e/import_test.go (hunk 7 above)
Comment (richardcase) [reply test/e2e/import_test.go:62]: Now this works we should:
- get the kubeconfig for the child cluster and check that we can connect.
- Check the status of the Rancher Cluster

File: test/go.mod [hunk 13]
```diff
@@ -0,0 +1,130 @@
+module github.com/rancher-sandbox/rancher-turtles/test
+
+go 1.20
+
+replace github.com/rancher-sandbox/rancher-turtles => ../
+
+replace sigs.k8s.io/cluster-api-operator/test => github.com/Danil-Grigorev/cluster-api-operator/test v0.0.0-20230814131416-1045122a87bf
```
Comment (richardcase) [inline test/go.mod:7]: Why are we doing the replace to your fork of CAPI operator?

**Review by richardcase — COMMENTED** [review]
Looking good, great work @Danil-Grigorev

File: test/go.mod (hunk 13 above)
Comment (furkatgofurov7) [reply test/go.mod:7]: Good catch, perhaps before changes landing in capi operator this was needed

File: hack/utils.sh [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+# get_capd_root_path returns the root path of CAPD source tree
+get_capd_root_path() {
+    echo "$(get_root_path)"/test/infrastructure/docker
+}
+
+# cd_capd_root_path cds to the root path of the CAPD source tree
+cd_capd_root_path() {
+    cd "$(get_capd_root_path)" || exit
+}
```
Comment (furkatgofurov7) [inline hack/utils.sh:34]: I am not sure how this would work, consider path `test/infrastructure/docker` does not exist

Comment (Danil-Grigorev) [reply hack/utils.sh:34]: This one specifically is not used anywhere, will remove it. Only get_root_path is important in the context

File: test/e2e/e2e_suite_test.go [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+	// usePRArtifacts specifies whether or not to use the build from a PR of the Kubernetes repository.
+	usePRArtifacts bool
+
+	// helmChart is the helm chart helper to be used for the e2e tests.
+	helmChart *HelmChart
+)
+
+func init() {
+	flag.StringVar(&configPath, "e2e.config", "", "path to the e2e config file")
+	flag.StringVar(&artifactFolder, "e2e.artifacts-folder", "", "folder where e2e test artifact should be stored")
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply test/e2e/e2e_suite_test.go]: Added a default folder `_artifacts` here

File: test/go.mod (hunk 13 above)
Comment (Danil-Grigorev) [reply test/go.mod]: Fixed

File: test/e2e/resources/fleet-capi-test.yaml (hunk 6 above)
Comment (Danil-Grigorev) [reply test/e2e/resources/fleet-capi-test.yaml]: For now I’ve only added fleet scenario, and it has labels included by default. We can look at it later.

File: Makefile (hunk 1 above)
Comment (Danil-Grigorev) [reply Makefile:403]: I think this one could be closed.

File: test/e2e/resources/testdata/fleet-capi-test.yaml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+
+kind: GitRepo
+apiVersion: fleet.cattle.io/v1alpha1
+metadata:
+  name: manifests
+  namespace: fleet-local
+spec:
+  repo: https://github.com/Danil-Grigorev/fleet-clusters
```
Comment (furkatgofurov7) [inline test/e2e/resources/testdata/fleet-capi-test.yaml:8]: is this a leftover from local tests?

File: test/e2e/resources/testdata/fleet-capi-test.yaml [hunk 17]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (furkatgofurov7) [inline test/e2e/resources/testdata/fleet-capi-test.yaml:1]: nit: extra line

File: hack/ensure-kubectl.sh [hunk 18]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply hack/ensure-kubectl.sh:3]: this is not resolved yet

**Review by furkatgofurov7 — COMMENTED** [review]
General question: I don't see the GH actions being introduced in this PR, what are our plans for running e2e tests in CI, just locally?

**Comment by Danil-Grigorev** [comment]
> General question: I don't see the GH actions being introduced in this PR, what are our plans for running e2e tests in CI, just locally?

It is not in the scope of this PR, we can look at it later

**Review by richardcase — APPROVED** [review]
It would be good to squash the commits so they tell a story.

**Comment by furkatgofurov7** [comment]
> It would be good to squash the commits so they tell a story.

@Danil-Grigorev can you address this comment before we take this in? Otherwise this LGTM

**Comment by Danil-Grigorev** [comment]
@richardcase @furkatgofurov7 @salasberryfin I’ve addressed all outstanding comments, commits are squashed. Additionally there is an updated [README.md](https://github.com/rancher-sandbox/rancher-turtles/blob/704a67f30216e90c7332eb79c02a1991afb5ce1d/test/e2e/README.md) if you have any question to the approach.

File: test/e2e/README.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+  CAPI_INFRASTRUCTURE: "docker" # Default list of capi providers installed in the cluster. Using docker:latest by default. Could be expanded with `docker,azure` to include lates azure provider for example.
+  NGROK_API_KEY: "" # Key and token values for establishing ingress
+  NGROK_AUTHTOKEN: ""
+```
+
+## Testdata
+
+For simplicity each test case testdata is stored under `test/e2e/resources/testdata`. This allows to use it with golang [embed](https://pkg.go.dev/embed) to avoid error checks and mistakes in the path resolution while loading the resources.
+
+Import of the resources could be found in `test/e2e/helpers_test.go`.
```
Comment (furkatgofurov7) [inline test/e2e/README.md:62]:
```suggestion
Import of the resources can be found in `test/e2e/helpers_test.go`.
```

**Review by furkatgofurov7 — COMMENTED** [review]
No blockers, mostly nits/suggestions and clarification questions:

File: Makefile [hunk 20]
```diff
@@ -76,10 +100,19 @@ GO_APIDIFF_BIN := go-apidiff
...
+GINKGO_BIN := ginkgo
+GINGKO_VER := $(call get_go_version,github.com/onsi/ginkgo/v2)
+GINKGO := $(abspath $(TOOLS_BIN_DIR)/$(GINKGO_BIN)-$(GINGKO_VER))
+GINKGO_PKG := github.com/onsi/ginkgo/v2/ginkgo
+
 HELM_VER := v3.8.1
 HELM_BIN := helm
 HELM := $(TOOLS_BIN_DIR)/$(HELM_BIN)-$(HELM_VER)
 
+CAPI_OPERATOR_VER := 0.5.0
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply Makefile:112]: No I noticed no such issue with docker provider.

File: Makefile [hunk 21]
```diff
@@ -441,6 +488,23 @@ release-chart: $(HELM) $(KUSTOMIZE) $(RELEASE_DIR) $(CHART_DIR) $(CHART_PACKAGE_
 update-helm-repo:
 	./hack/update-helm-repo.sh $(RELEASE_TAG)
 
+.PHONY: test-e2e
+test-e2e: $(GINKGO) $(HELM) $(CAPI_OPERATOR) kubectl ## Run the end-to-end tests
+	RELEASE_TAG=v0.0.1 $(MAKE) docker-build
+	RELEASE_TAG=v0.0.1 $(MAKE) release-chart
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply Makefile:494]: This here is required for the local helm chart build. This is not an actual release, don’t see a point of parametrization here. The change was introduced based on feedback from @salasberryfin, as for him there was no initial tag in the repo, and a constant override here is needed.

File: internal/controllers/import_controller.go [hunk 22]
```diff
@@ -97,7 +95,7 @@ func (r *CAPIImportReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Ma
 
 	ns := &corev1.Namespace{}
 	err = c.Watch(
-		source.Kind(r.cache, ns),
+		source.Kind(mgr.GetCache(), ns),
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply internal/controllers/import_controller.go:98]: Maybe just missing rebase, will have a look

File: test/e2e/import_test.go [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+var _ = Describe("Create and delete CAPI cluster functionality should work", func() {
+	var (
+		rancherClusterHandler *rancher.ClusterHandler
+		rancherCluster        *rancher.Cluster
+		capiCluster           *clusterv1.Cluster
+		rancherClusterKey     client.ObjectKey
+		fleetGeneration       int
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply test/e2e/import_test.go:46]: When you remove the underlying resources provisioned by fleet, they don’t get re-created automatically. This forcefully applies a higher generation to force re-rollout. https://github.com/rancher/fleet/issues/163 describes it. Might be subject for removal after release with fix for https://github.com/rancher/fleet/issues/1551, forceSyncGeneration works fine with 2.7.5

File: hack/ensure-kubectl.sh (hunk 18 above)
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply hack/ensure-kubectl.sh]: Squashing commits is the reason for such problems. This was fixed, now twice.

File: internal/controllers/import_controller.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply internal/controllers/import_controller.go]: Rebased

File: test/e2e/import_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply test/e2e/import_test.go:46]: In that case, should we leave a note with TODO capturing this so we don't forget to remove it once it is fixed?

[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply test/e2e/import_test.go]: Added

File: hack/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml [hunk 24]
```diff
@@ -0,0 +1,1142 @@
+---
```
Comment (furkatgofurov7) [inline hack/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml:1]: One last question: why we have these crds generated now since I last time reviewed, what has changed?

Comment (Danil-Grigorev) [reply hack/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml]: Sorry, due to switching between multiple commits/PRs I accidentally added what is gitignored in https://github.com/rancher-sandbox/rancher-turtles/pull/84 and allows us to support structured types and run tests. Fixrd

**Review by furkatgofurov7 — DISMISSED** [review]
Thanks @Danil-Grigorev, this version looks good to me and we can iterate over if needed.

/approve

**Review by alexander-demicev — DISMISSED** [review]
great work! we can merge it after rebase

**Review by richardcase — APPROVED** [review]
Looks good to me as well, great work @Danil-Grigorev

**Review by furkatgofurov7 — APPROVED** [review]

**Review by alexander-demicev — APPROVED** [review]
